integration/run_test.go

```diff
@@ -296,135 +286,80 @@ max_wait_msec = ` + strconv.FormatInt(tt.config.maxWaitMsec, 10) + `
+			rebootContainerd(t, sh, getContainerdConfigToml(t, false), getSnapshotterConfigToml(t, false, config))
 			// Mirror image
 			copyImage(sh, dockerhub(containerImage), regConfig.mirror(containerImage))
-			// Pull image, create SOCI index with all layers and small (100kiB) spans
-			indexDigest := buildIndex(sh, regConfig.mirror(containerImage), withMinLayerSize(0), withSpanSize(100*1024))
-			sh.X(append(imagePullCmd, "--soci-index-digest", indexDigest, regConfig.mirror(containerImage).ref)...)
-
-			// Run the container
 			image := regConfig.mirror(containerImage).ref
-			sh.X(append(runSociCmd, "--name", "test-container", "-d", image, "sleep", "infinity")...)
-
-			sh.X("apt-get", "--no-install-recommends", "install", "-qy", "iptables")
-
-			// TODO: Wait for the container to be up and running
+			indexDigest := buildIndex(sh, regConfig.mirror(containerImage), withMinLayerSize(0), withSpanSize(1<<20))
+			sh.X("soci", "push", "--user", regConfig.creds(), regConfig.mirror(containerImage).ref)
 
+			rebootContainerd(t, sh, getContainerdConfigToml(t, false), getSnapshotterConfigToml(t, false, config))
+			// Re-pull image from our local registry mirror
+			sh.X(append(imagePullCmd, "--soci-index-digest", indexDigest, regConfig.mirror(containerImage).ref)...)
+			sh.X("apt-get", "-qq", "--no-install-recommends", "install", "-y", "iptables")
 			// Block network access to the registry
 			if tt.config.networkDisableMsec > 0 {
 				sh.X("iptables", "-A", "OUTPUT", "-d", registryHostIP, "-j", "DROP")
 			}
 
-			// Do something in the container that should work without network access
-			commandSucceedStdout, _, err := sh.R("nerdctl", "exec", "test-container", "sh", "-c", "times")
+			// Read a file to trigger a synchronous network request.
+			cmdStdout, cmdStderr, err := sh.R(append(runSociCmd, image, "cat", "/etc/hosts")...)
 			if err != nil {
-				t.Fatalf("attempt to run task without network access failed: %s", err)
-			}
-
-			type ErrorLogLine struct {
-				Error string `json:"error"`
-				Msg   string `json:"msg"`
+				t.Fatalf("attempt to run task requiring network access failed: %s", err)
 			}
 
-			gaveUpChannel := make(chan bool, 1)
-			defer close(gaveUpChannel)
+			successChannel := make(chan string, 1)
+			failureChannel := make(chan string, 1)
 
-			monitorGaveUp := func(rawL string) {
-				if i := strings.Index(rawL, "{"); i > 0 {
-					rawL = rawL[i:] // trim garbage chars; expects "{...}"-styled JSON log
-				}
-				var logLine ErrorLogLine
-				if err := json.Unmarshal([]byte(rawL), &logLine); err == nil {
-					if logLine.Msg == "statFile error" && strings.Contains(logLine.Error, "giving up after") {
-						gaveUpChannel <- true
+			listener := func(c chan string) func(string) {
+				return func(rawLog string) {
+					select { // Make sure this isn't blocking
+					case c <- rawLog:
+					default:
 						return
 					}
```

Comment:
Let's include a comment here to explain more of why we don't want to block. i.e. the channel is unbuffered and we only really care about the first message so write the message to the channel and drop everything else.

Alternatively it might be more straightforward to have a `sync.Once` for writing to the channel and close the channel when you have written the first message.

Reply:
WDYT about this?

```golang
			listener := func(c chan string) func(string) {
				return func(rawLog string) {
					var once sync.Once
					tempFunc := func() {
						c <- rawLog
					}
					once.Do(tempFunc)
				}
			}
```